### ISSUE TITLE
When to implement abstract class?

### ISSUE THREAD

**Issue by NWH21** [nus-cs2030/2122-s2#41]
I have submitted my solution for lab 4 which included a service class that was a parent class for the various cab services provided. The service class did not have any constructors and upon submission did I then realize that actually I could have submitted an abstract class instead, which in this case would have seemed to be more appropriate. I was just wondering if there are any particular conditions for us to look out for as to when to implement an abstract class?

**Comment by jonghyunsohn** [comment]
Yeah I also came up on the same realization when I was working on my JustRide and TakeACab code. I think cases where you can create one abstract code (like the Service code for Lab 4) and extend different calculations in other classes (like JustRide and TakeACab) is the best time to use it.

**Comment by hongyaong** [comment]
Hi! Good question! I was also wondering the same thing. I have used abstract class "service" as the parent and the services as child, but only had an abstract method which all classes will Override differently. But at the same time, I heard it can be made such that it is a concrete class with a method that is smart enough to calculate for different service. haha. And from online sources, some conditions to keep in mind is that Abstract classes cannot be instantiated and abstract methods are used for overriding purpose. I guess use abstract class when we want to create abstract methods for child?

**Comment by Darien2307** [comment]
Hi guys! For me, I felt like because the fare computation was very different across various services (and possibly other services that can be implemented but not within the scope of the question), I decided to just go ahead with an interface for Service. Not too sure in this case, what's the significant difference between using an abstract class or interface? In the case for a Driver class, I decided to go with abstract class because different drivers offer different services so I would like to include a serviceList in this driver class that can be overriden and obtain the correct services offered by different drivers. However, it wouldn't make sense to return an empty list in the main Driver class also. Then my thought would be not to make it a normal class. Additionally, for compareTo() in Booking class, if fare is the same, waiting time will be the next factor for consideration. So to obtain the waiting time, i will have to access it from the driver classes (NormalCab and PrivateCar). But then because the attributes have to be private final, I will therefore need get methods. Since get() will be applicable to all drivers, it can be listed as public methods in the Driver class. In conclusion, an abstract class seems necessary for Driver. I'm not sure if this is a reasonable thought process also...Maybe someone can advise?

**Comment by shotnothing** [comment]
Hi, I am not too certain but my understanding was that it is a matter of preference and discretion, where interfaces are simpler to implement but abstract classes are more useful if there is functionality intrinsic among all the child classes and/or the parent makes sense to exist by itself.

E.g. 
A square is a special type of rectangle, but a rectangle can exist alone. Methods like getArea and getPerimeter work for both classes, so by implementing them in rectangle, they would work for both.

A driver, however, is useless if it exists alone (abstract class). It cannot "promise" methods that call Service::computeFare since the base driver class does not have any services. 

If driver has a method bestService() that calls Service::computeFare, and you were to do

```
for driver in list of drivers
       driver.bestService()
```

then there is a chance that if the client manages to pass a driver from the base driver class, the code will break.

You could always try and make sure that the base driver class is never added to the list, but it seems better practice and logical sense to implement it as an interface. Thus, I think its a matter of preference and discretion which is used.

**Comment by weekiatism** [comment]
I would say there is a degree of flexibility to how you implement your solution. However, take note that abstract classes cannot be instantiated on its own. Hence, the purpose is to provide a template for their child classes.

I feel that abstract classes are suitable for scenarios where multiple child classes have the methods in common, with a mix of a few which are identical and a few which contents are unique to the child class. 

Abstract classes is an in between implementation of both interfaces and classes:

For interfaces, it is suitable if there are methods in the child class in common, but with different contents (e.g diff way of calculating fare). But the method contents are unfixed.

For classes, it is suitable if there are methods in the child class in common with the same contents (e.g same way of doing something).
